### DISCUSSION TITLE
Does SeaweedFS support upload via presigned url? (Browser-based upload using HTTP POST)

### DISCUSSION THREAD

**Discussion by daghendrik** [seaweedfs/seaweedfs#4401] [Q&A]
Is it possible to use the AWS SDK (for example in Java) to create a pre-signed url for uploading to SeaweedFS's S3 gateway? 

I tried, but I get a 405 (Method Not Allowed) when performing the POST request to the localhost S3 endpoint (http://127.0.0.1:8333/) from my webapp (localhost angular app).

AWS DOCS:

- https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-post-example.html
- https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-UsingHTTPPOST.html

Similar to MinIO: https://min.io/docs/minio/linux/integrations/presigned-put-upload-via-browser.html

**Comment by spirobel** [comment]
same here. Did you find a solution? I am still on the hunt ... currently investigating adding a bucket policy that allows the method ...

**Reply by daghendrik** [reply]
Unfortunately, no. I haven't looked at SeaweedFS since April 2023, so this might have been added later on.

**Reply by spirobel** [reply]
I also tried garagehq after running into the 405 post method not allowed cors problem. But there I ran into a new problem: ><Code>AccessDenied</Code><Message>Forbidden: Garage does not support anonymous access yet</Message>

so it seems like for this use case seaweedfs does not work and garagehq is a deadend as well.

Did you have better luck with minio?

**Reply by daghendrik** [reply]
Yes, with minio it worked without any additional configurations

**Comment by kmlebedev** [comment]
There is full support for presigned requests, but you need someone to actually use it in order to send a PR and be able to debug it.

**Reply by spirobel** [reply]
https://zaccharles.medium.com/s3-uploads-proxies-vs-presigned-urls-vs-presigned-posts-9661e2b37932
I followed this tutorial. But I was not able to set POST in the allowedmethod cors settings. And the bucket cors policy is not supported by seaweedfs. (i looked at the code after trying to set it with the aws client sdk)

**Reply by spirobel** [reply]
anyway thanks for your hard work and all the best with seaweedfs!

**Reply by qiangswa** [reply]
signed url for putObject works @spirobel  with following config. What I found is that you have to specify ContentType at both backed (while signing) and UI (while uploading)

 const params = {
      Bucket: this.bucket,
      Key: key,
      ContentType: 'application/octet-stream',  // need this to make seaweedfs work
      Expires: 3600, // URL expires in 1 hour (you can adjust this value as needed)
    };

for getObject it works with following params

 const params = {
      Bucket: this.bucket,
      Key: key,
      Expires: 3600, // URL expires in 1 hour (you can adjust this value as needed)
    };

**Comment by nikola-s7** [comment]
Apologies for resurrecting an old thread, but I just ran into this very issue and have been banging my head against it for the last two days.

I just thought I'd share my solution for anyone having the same problem in the future.

In my specific case I was using the package `@aws-sdk/s3-presigned-post`, specifically its `createPresignedPost` function to generate the url and form data for the POST request.

When using the URL return by `createPresignedPost` (which has the form `<BUCKET>.<SEAWEEDFS_DOMAIN>:<SEAWEEDFS_PORT>/`), SeaweedFS returns a 405 Method Not Allowed response.

It turns out, that the correct URL to use moves the bucket name from the subdomain to the path, i.e. `<SEAWEEDFS_DOMAIN>:<SEAWEEDFS_PORT>/<BUCKET>`. This was my main issue and should everything work for you from here, congrats.


You might however, get a 403 Forbidden response with the error message along the lines of
```
<Error>
  <Code>InvalidAccessKeyId</Code>
  <Message>The access key ID you provided does not exist in our records.</Message>
  <Resource>/bucketname</Resource>
  <RequestId>1234567849</RequestId>
  <BucketName>bucketname</BucketName>
</Error>
```

In this case, you need to ensure that you have configured an identity with proper permission and are using the correct credentials (accessKey, secretKey) with the S3Client.

**Reply by william1benn** [reply]
Too add to this as I came across this same issue, the easiest way to resolve to avoid manually editing urls when using the aws sdk is to just put the **BUCKET_NAME** as part of the **OBJECT_KEY**  to specify the the full path in the s3 presign request configuration. 

**Below is sdk golang example** 

```
	presignUpload, err := presignClient.PresignPutObject(context.TODO(), &s3.PutObjectInput{
		Bucket:      *"BUCKET_NAME",
		Key:       *"BUCKETNAME/OBJECT_KEY",
	})
```

_If anyone has a general fix, instead of this work around please share_.
